cmd/minishift/cmd/profile/profile_copy.go

```diff
@@ -0,0 +1,142 @@
+	}
+
+	// Get the config file of Profile from which we need to copy the config
+	srcConfFile := constants.GetProfileConfigFile(srcProfile)
+
+	// Read the config from the profile we need to copy it
+	sourceConfig, err := config.ReadViperConfig(srcConfFile)
+	if err != nil {
+		atexit.ExitWithMessage(1, err.Error())
+	}
+
+	//Create the directory structure for the new profile
+	newInstanceDirs := state.GetMinishiftDirsStructure(constants.GetProfileHomeDir(profileName))
+	cmdUtil.CreateMinishiftDirs(newInstanceDirs)
+
+	// Write the config information to the new configfile
+	targetConfFile := constants.GetProfileConfigFile(profileName)
+
+	//Create the config.json for the new profile
+	cmdUtil.EnsureConfigFileExists(targetConfFile)
+
+	// Check if the config file for the new profile is not empty
+	targetConfig, err := config.ReadViperConfig(targetConfFile)
+	if err != nil {
+		atexit.ExitWithMessage(1, err.Error())
+	}
+	if len(targetConfig) != 0 {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Can not copy configurations from '%s' to '%s' as '%s' is not empty", srcProfile, profileName, targetConfFile))
+	}
+
+	err = config.WriteViperConfig(targetConfFile, sourceConfig)
+	if err != nil {
+		atexit.ExitWithMessage(1, err.Error())
+	}
+
+	//Get directory structure of the source profile
+	srcProfileDirs := cmdState.GetMinishiftDirsStructure(constants.GetProfileHomeDir(srcProfile))
+
+	// Copy addons from source profile to new profile
+	// Removing the newly created addon directory as filehelper.CopyDir does not expect the directory to be present.
+	err = os.Remove(newInstanceDirs.Addons)
+	if err != nil {
+		atexit.ExitWithMessage(1, err.Error())
+	}
+
+	err = filehelper.CopyDir(srcProfileDirs.Addons, newInstanceDirs.Addons)
+	if err != nil {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Copying add-ons to profile '%s' failed : %s", profileName, err.Error()))
+	}
+
+	// Get the machine config file (MACHINE_NAME.json) of the target profile
+	targetProfileInstanceConfig, err := config.NewInstanceConfig(minishiftConstants.GetProfileInstanceConfigPath(profileName))
+	if err != nil {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Error creating config for profile '%s' : %s", profileName, err.Error()))
+	}
+
+	//Get the instance machine config of the source profile
+	srcProfileInstanceConfig, err := config.NewInstanceConfig(minishiftConstants.GetProfileInstanceConfigPath(srcProfile))
+	if err != nil {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Error creating config for source profile '%s' : %s", srcProfile, err.Error()))
+	}
+
+	// Copy addon state information from config/minishift.json
+	targetProfileInstanceConfig.AddonConfig = srcProfileInstanceConfig.AddonConfig
+	if err != targetProfileInstanceConfig.Write() {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Error copying add-on config to profile '%s': %s", profileName, err.Error()))
+	}
+
+	fmt.Println(fmt.Sprintf("Profile '%s' is created successfully using configs from profile '%s'", profileName, srcProfile))
+}
```

Comment:
Why don't you put this logic to a separate function and then you can also put some unit tests around it easily.

```golang
func copySrcToDstProfile(srcProfile, profileName string) {

	// Get the config file of Profile from which we need to copy the config
	srcConfFile := constants.GetProfileConfigFile(srcProfile)
	// Read the config from the profile we need to copy it
	sourceConfig, err := config.ReadViperConfig(srcConfFile)
	if err != nil {
		atexit.ExitWithMessage(1, err.Error())
	}

	//Create the directory structure for the new profile
	newInstanceDirs := state.GetMinishiftDirsStructure(constants.GetProfileHomeDir(profileName))
	cmdUtil.CreateMinishiftDirs(newInstanceDirs)

	// Write the config information to the new configfile
	targetConfFile := constants.GetProfileConfigFile(profileName)

	//Create the config.json for the new profile
	cmdUtil.EnsureConfigFileExists(targetConfFile)

	// Check if the config file for the new profile is not empty
	targetConfig, err := config.ReadViperConfig(targetConfFile)
	if err != nil {
		atexit.ExitWithMessage(1, err.Error())
	}
	if len(targetConfig) != 0 {
		atexit.ExitWithMessage(1, fmt.Sprintf("Can not copy configurations from '%s' to '%s' as '%s' is not empty", srcProfile, profileName, targetConfFile))
	}

	err = config.WriteViperConfig(targetConfFile, sourceConfig)
	if err != nil {
		atexit.ExitWithMessage(1, err.Error())
	}

	//Get directory structure of the source profile
	srcProfileDirs := cmdState.GetMinishiftDirsStructure(constants.GetProfileHomeDir(srcProfile))

	// Copy addons from source profile to new profile
	// Removing the newly created addon directory as filehelper.CopyDir does not expect the directory to be present.
	err = os.Remove(newInstanceDirs.Addons)
	if err != nil {
		atexit.ExitWithMessage(1, err.Error())
	}

	err = filehelper.CopyDir(srcProfileDirs.Addons, newInstanceDirs.Addons)
	if err != nil {
		atexit.ExitWithMessage(1, fmt.Sprintf("Copying add-ons to profile '%s' failed : %s", profileName, err.Error()))
	}

	// Get the machine config file (MACHINE_NAME.json) of the target profile
	targetProfileInstanceConfig, err := config.NewInstanceConfig(minishiftConstants.GetProfileInstanceConfigPath(profileName))
	if err != nil {
		atexit.ExitWithMessage(1, fmt.Sprintf("Error creating config for profile '%s' : %s", profileName, err.Error()))
	}

	//Get the instance machine config of the source profile
	srcProfileInstanceConfig, err := config.NewInstanceConfig(minishiftConstants.GetProfileInstanceConfigPath(srcProfile))
	if err != nil {
		atexit.ExitWithMessage(1, fmt.Sprintf("Error creating config for source profile '%s' : %s", srcProfile, err.Error()))
	}

	// Copy addon state information from config/minishift.json
	targetProfileInstanceConfig.AddonConfig = srcProfileInstanceConfig.AddonConfig
	if err != targetProfileInstanceConfig.Write() {
		atexit.ExitWithMessage(1, fmt.Sprintf("Error copying add-on config to profile '%s': %s", profileName, err.Error()))
	}

	fmt.Println(fmt.Sprintf("Profile '%s' is created successfully using configs from profile '%s'", profileName, srcProfile))
}
```

Reply:
@praveenkumar There actual files involved here, so I can only write tests when I can mock creating a profile

Reply:
Mocking of the profile is quite easy I think, can you try something like following and see if that works.

```golang
dir, err := ioutil.TempDir("", "minishift-home")
if err != nil {
	log.Fatal(err)
}
defer os.RemoveAll(dir)
os.Setenv("MINISHIFT_HOME", dir)
[Now what ever profile or file we create should be part of tmpdir]
```

Reply:
Is something possible here, such as extraction to separate named functions? WDYT

Reply:
So this is accepted?